addons/hr_attendance/tests/test_attendance_process.py

```diff
@@ -0,0 +1,110 @@
+# -*- coding: utf-8 -*-
+
+import time
+
+from openerp.exceptions import ValidationError
+from openerp.tests import common
+
+
+class TestAttendanceProcess(common.TransactionCase):
+
+    def setUp(self):
+        super(TestAttendanceProcess, self).setUp()
+
+        self.ResUsers = self.env['res.users']
+        self.ResEmployee = self.env['hr.employee']
+        self.Attendance = self.env['hr.attendance']
+        self.hr_employee_niv = self.ref('hr.employee_niv')
+        # Create a user as 'HR Attendance Officer.'
+        self.res_users_attendance_officer = self.ResUsers.create({
+            'company_id': self.ref('base.main_company'),
+            'name': 'HR Officer',
+            'login': 'ao',
+            'password': 'ao',
+            'groups_id': [(6, 0, [self.ref('base.group_hr_user')])]})
+        self.hr_employe_all = self.env.ref('hr.employee_al').with_context({'uid': self.res_users_attendance_officer.id})
+
+    def test_hr_attendance_process(self):
+        # Give the access rights of Hr Officer to test attendance process.
+        # In order to test attendance process in OpenERP, I entry of signin of employee.
+
+        self.hr_employe_all.attendance_action_change()
+        #-------------------------------------
+        # I check that employee is "present".
+        #-------------------------------------
+
+        self.hr_employe_all.invalidate_cache()
+        self.assertEqual(self.hr_employe_all.state, 'present', 'Employee should be in present state.')
+        # After few seconds, employee sign's out.
+        time.sleep(2)
+        self.hr_employe_all.attendance_action_change()
+
+        #-------------------------------------
+        # I check that employee is "absent".
+        #-------------------------------------
+        self.hr_employe_all.invalidate_cache()
```

Comment:
@msh : when I called attendance_action_change method to SIGN IN. first value is Sign_in. and it stores in cache . to clear cache for next action,  we have to you use invalidate_cache() method over here.